applications/server/modules/access.py

```diff
@@ -0,0 +1,83 @@
+from gluon import current
+from gluon import auth
+
+
+# Permission types.
+# v = view
+# a = admin (valid only for one whole device)
+# e = edit settings of procedure
+
+def register_device(name=None, description=''):
+    db = current.db
+    devices = db.devices
+    permission = db.user_permission
+    device_id = devices.insert(name=name, user_id=auth.user_id, description=description)
+    permission.insert(perm_user_email=auth.user.email, device_id=device_id, procedure_id=None, perm_type='a')
+
+
+# Share one device to multiple users with one permission type
+# return emails which are not successfully shared.
+def share_device(device_id, user_email=[], type='v', procedure_id=None):
+    db = current.db
+    permission = db.user_permission
+    fail_email = []
+    for email in user_email:
+        # Does the email exist in registered user table?
+        # if it does not exist this permission adding operation is denied
```

Comment:
Yes, I agree with you. Correct me if I was wrong, if a user does not register on our website, we can send an invitation to them. There are two plans come to my mind, one is to send out the invitations directly, the second one is to notice the user that these people he can invite then let the user decide whether to invite. If we add the permissions to our table immediately, it may cause inconsistency because these users do not exist in auth_user table?